gh-114314: ctypes: remove stgdict and switch to heap types

encukou opened the pull request:
Before this change, ctypes classes used a custom dict subclass, `StgDict`,
as their `tp_dict`. This acts like a regular dict but also includes extra information
about the type.

This replaces stgdict by `StgInfo`, a C struct on the type, accessed by
`PyObject_GetTypeData()` (PEP-697).
All usage of `StgDict` (mainly variables named `stgdict`, `dict`, `edict` etc.) is
converted to `StgInfo` (named `stginfo`, `info`, `einfo`, etc.).
Where the dict is actually used for class attributes (as a regular PyDict), it's now
called `attrdict`.

This change -- not overriding `tp_dict` -- is made to make me comfortable with
the next part of this PR: moving the initialization logic from `tp_new` to `tp_init`.

The `StgInfo` is set up in `__init__` of each class, with a guard that prevents
calling `__init__` more than once. Note that abstract classes (like `Array` or
`Structure`) are created using `PyType_FromMetaclass` and do not have
`__init__` called.
Previously, this was done in `__new__`, which also wasn't called for abstract
classes.
Since `__init__` can be called from Python code or skipped, there is a tested
guard to ensure `StgInfo` is initialized exactly once before it's used.

Thanks to @neonene and @erlend-aasland.

<!-- gh-issue-number: gh-114314 -->
* Issue: gh-114314
<!-- /gh-issue-number -->

encukou committed:
Move paramfunc into a new struct, PyStgInfo

In this bootstrap commit, PyStgInfo is part of StgDictObject,
but that's a detail hidden by accessor functions.

(In the accessors, `state` is currently unused.)

encukou committed:
Move the StgInfo out of StgDictObject and into the classes themselves

encukou committed:
Move StgInfo init/retrieval right after the corresponding StgDict one

encukou committed:
Move `size` from StgDict to StgInfo

encukou committed:
Move `align` from StgDict to StgInfo

encukou committed:
Move `length` from StgDict to StgInfo, except for sizeof calculation

encukou committed:
Move part of the sizeof calculation

encukou committed:
Move `ffi_type_pointer` from StgDict to StgInfo

encukou committed:
Move `proto` from StgDict to StgInfo. Doesn't work just yet.

encukou committed:
Hack: Reserve space for StgInfo in static type objects

encukou committed:
If StgInfo is not initialized, act as if it's not there

encukou committed:
Move `getfunc` & `setfunc`

encukou committed:
Move `argtypes`

encukou committed:
Move `converters`

encukou committed:
Move `restype`

encukou committed:
Move `checker`

encukou committed:
Move `flags`

encukou committed:
Move `format`

encukou committed:
Move `ndim`

encukou committed:
Move `shape`

encukou committed:
Replace more uses of StgDict

encukou committed:
Don't pass dicts to PyCStgDict_clone

encukou committed:
Don't set StgDict as type dict; use name `attrdict` for the type __dict__

encukou committed:
Remove more uses of StgDict

encukou committed:
Remove StgDict

encukou committed:
Remove StgDict from comments, names, and platform-specific code

encukou committed:
Switch PyCSimpleType_new to _init

encukou committed:
Switch _SimpleCData to a heap type

encukou committed:
Convert Struct and Union type _new to _init

encukou committed:
Switch PyCPointerType_new to _init

encukou committed:
Switch PyCFuncPtrType_new to _init

encukou committed:
Switch CFuncPtr to heap type

encukou committed:
Switch Array to heap type

encukou committed:
Switch _Pointer to a heap type

encukou committed:
Switch _ctypes.Union to heap type

encukou committed:
Switch Structure to heap type

encukou committed:
Make the type-adding macros more regular

encukou committed:
Switch _CData to a heap type

encukou committed:
Visit/decref type in PyCData visit/dealloc

encukou committed:
Remove _HackyHeapType

encukou committed:
Decref attrdicts

encukou committed:
Delegate CType_Type deallocation to PyType_Type.tp_dealloc

neonene committed:
Fix Windows-specific crashes

encukou committed:
Update comments

encukou committed:
Add tests for metaclass details

encukou committed:
Fix getting the module state

encukou committed:
Add a blurb, even though this *should* not affect usage

erlend-aasland commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -459,20 +444,112 @@ static PyType_Spec structparam_spec = {
+static void
+CType_Type_dealloc(PyObject *self)
+{
+    ctypes_state *st = GLOBAL_STATE();
```
We've had some segfaults caused by module access in `tp_dealloc` recently. We should make sure this is really safe.

erlend-aasland commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -459,20 +444,112 @@ static PyType_Spec structparam_spec = {
+    ctypes_state *st = GLOBAL_STATE();
+    StgInfo *info;
+    if (PyStgInfo_FromType(st, self, &info) < 0) {
+        // ignore exception
```
Print unraisable exception?

erlend-aasland commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -459,20 +444,112 @@ static PyType_Spec structparam_spec = {
+    }
+
+    PyTypeObject *tp = Py_TYPE(self);
+    (void)CType_Type_clear(self);
```
Check error and print unraisable exception?

erlend-aasland commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -507,104 +584,103 @@ StructUnionType_paramfunc(CDataObject *self)
-        return (PyObject *)result;
+
+        /* copy base info */
+        if (-1 == PyCStgInfo_clone(info, baseinfo)) {
```
```suggestion
        if (PyCStgInfo_clone(info, baseinfo) < 0) {
```

erlend-aasland commented on Modules/_ctypes/callbacks.c:
```diff
@@ -154,22 +158,25 @@ static void _CallPythonObject(void *mem,
                BTW, the same problem occurs when they are pushed as parameters
             */
-        } else if (dict) {
+        } else if (info) {
```
```suggestion
        }
        else if (info) {
```

erlend-aasland commented on Modules/_ctypes/stgdict.c:
```diff
@@ -16,201 +16,121 @@
+}
+// from either a type or an instance:
+int
+PyStgInfo_FromAny(ctypes_state *state, PyObject *obj, StgInfo **result) {
```
```suggestion
PyStgInfo_FromAny(ctypes_state *state, PyObject *obj, StgInfo **result)
{
```

erlend-aasland commented on Modules/_ctypes/stgdict.c:
```diff
@@ -926,27 +864,36 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
+                Py_ssize_t length = info->length;
+                StgInfo *einfo;
+                if (PyStgInfo_FromType(st, info->proto, &einfo) < 0) {
+                    return -1;
```
```suggestion
                    Py_DECREF(pair);
                    PyMem_Free(type_block);
                    return -1;
```

erlend-aasland commented on Modules/_ctypes/stgdict.c:
```diff
@@ -829,25 +760,33 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
-                if (edict == NULL) {
+                StgInfo *einfo;
+                if (PyStgInfo_FromType(st, info->proto, &einfo) < 0) {
+                    return -1;
```
```suggestion
                    Py_DECREF(pair);
                    return -1;
```

erlend-aasland commented on Modules/_ctypes/stgdict.c:
```diff
@@ -16,201 +16,121 @@
+        // not a ctypes class.
+        return 0;
+    }
+    StgInfo *info = PyObject_GetTypeData((PyObject *)type, state->PyCType_Type);
```
According to the docs, `PyObject_GetTypeData` raises an exception if `NULL` is returned, however we know that the implementation (currently) does not raise an exception. It may be nice to future readers of the code to make this more explicit, for example by adding an assert.

```suggestion
    StgInfo *info = PyObject_GetTypeData((PyObject *)type, state->PyCType_Type);
    assert(info != NULL);
```

erlend-aasland commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -459,20 +444,112 @@ static PyType_Spec structparam_spec = {
+
+    PyTypeObject *tp = Py_TYPE(self);
+    (void)CType_Type_clear(self);
+    PyType_Type.tp_dealloc(self);
+    Py_DECREF(tp);
```
Also untrack from GC?

encukou replied:
`PyType_Type.tp_dealloc` will do that.

erlend-aasland commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -459,20 +444,112 @@ static PyType_Spec structparam_spec = {
+}
+
+static PyObject *
+CType_Type_sizeof(PyObject *self) {
```
```suggestion
CType_Type_sizeof(PyObject *self)
{
```

erlend-aasland reviewed (commented):
Some comments after a partial review. My impression is that this is a very nice solution to the problem; good job!

Regarding the internal `PyStgInfo_From*` APIs; these never fail with an exception (`-1` is never returned), yet all callers check for `if (func() < 0)`. For an external API, this makes sense, but is it really needed for a `_ctypes` internal API? The old StgDict API was designed to be as fast as possible. Could we not just `assert()` that the result is `>= 0` for these calls? Alternatively design the API so that we only return the `StgInfo` pointer or `NULL`.

I'm not halfway done reviewing `_ctypes.c` yet; I need more days for that one :)

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -459,20 +444,112 @@ static PyType_Spec structparam_spec = {
+
+static PyType_Slot ctype_type_slots[] = {
+    {Py_tp_traverse, CType_Type_traverse},
+    {Py_tp_clear, CType_Type_clear},
```
```suggestion
```
Previously, `PyCStgDict_clear()` was not a `tp_clear` function.  `CType_Type_clear()` also seems to work just for `CType_Type_dealloc()` (mixable).
https://github.com/python/cpython/blob/7cee276d551a41d9271daf2a6bcd7def55555973/Modules/_ctypes/stgdict.c#L161-L162

neonene replied:
Alternatively, it would be less confusing to integrate `CDataType_*()` into `CType_Type_*()`, keeping the `{Py_tp_clear, CType_Type_clear}` slot for subtypes to inherit, roughly like the following:

<details>

```diff
 static int
 CType_Type_traverse(PyObject *self, visitproc visit, void *arg)
 {
+    ctypes_state *st = GLOBAL_STATE();
+    StgInfo *info;
+    if (PyStgInfo_FromType(st, self, &info) < 0) {
+        return -1;
+    }
+    if (info) {
+        Py_VISIT(info->proto);    // GC-related
+        Py_VISIT(info->argtypes);
+        Py_VISIT(info->converters);
+        Py_VISIT(info->restype);
+        Py_VISIT(info->checker);
+    }
     Py_VISIT(Py_TYPE(self));
-    return 0;
+    return PyType_Type.tp_traverse(self, visit, arg);
 }

 static int
 CType_Type_clear(PyObject *self)
 {
     ctypes_state *st = GLOBAL_STATE();
     StgInfo *info;
     if (PyStgInfo_FromType(st, self, &info) < 0) {
         return -1;
     }
     if (info) {
         Py_CLEAR(info->proto);    // GC-related
         Py_CLEAR(info->argtypes);
         Py_CLEAR(info->converters);
         Py_CLEAR(info->restype);
         Py_CLEAR(info->checker);
     }
-    return 0;
+    return PyType_Type.tp_clear(self);
 }

 static PyType_Slot foo_type_slots[] = {
     ...
-    {Py_tp_traverse, CDataType_traverse},
-    {Py_tp_clear, CDataType_clear},
     ...
 };

 PyType_Spec foo_type_spec = {
     ...
-    .flags = (Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC |
+    .flags = (Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE |
     ...
 };
```
</details>

encukou replied:
That would indeed be clearer! I'd prefer leaving it to a follow-up PR though: it'll be easier to debug when we get to isolating the module state.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -507,104 +584,103 @@ StructUnionType_paramfunc(CDataObject *self)
-        Py_DECREF(result);
-        return NULL;
+    if (PyDict_GetItemRef((PyObject *)attrdict, &_Py_ID(_fields_), &fields) < 0) {
+        return -1;
```
```suggestion
        Py_DECREF(attrdict);
        return -1;
```

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2006,70 +2099,61 @@ PyCSimpleType_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
-    if (stgdict->format == NULL) {
-        Py_DECREF(result);
+    if (stginfo->format == NULL) {
+        Py_DECREF(self);
```
```suggestion
```
IIUC, [`type_call()`](https://github.com/python/cpython/blob/7cee276d551a41d9271daf2a6bcd7def55555973/Objects/typeobject.c#L1788-L1793) in `typeobject.c` does this.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2079,57 +2163,62 @@ PyCSimpleType_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
             if (!meth) {
-                Py_DECREF(result);
-                return NULL;
+                Py_DECREF(self);
```
```suggestion
```
Ditto.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2079,57 +2163,62 @@ PyCSimpleType_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
             if (x == -1) {
-                Py_DECREF(result);
-                return NULL;
+                Py_DECREF(self);
```
```suggestion
```
Ditto.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2079,57 +2163,62 @@ PyCSimpleType_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
         if (swapped == NULL) {
-            Py_DECREF(result);
-            return NULL;
+            Py_DECREF(self);
```
```suggestion
```
Ditto.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2079,57 +2163,62 @@ PyCSimpleType_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+        }
+        StgInfo *sw_info;
+        if (PyStgInfo_FromType(st, swapped, &sw_info) < 0) {
+            Py_DECREF(self);
```
```suggestion
```
Ditto.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2079,57 +2163,62 @@ PyCSimpleType_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 #endif
         Py_DECREF(swapped);
         if (PyErr_Occurred()) {
             Py_DECREF(result);
```
```suggestion
```
Ditto.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -1948,33 +2047,27 @@ PyCSimpleType_paramfunc(CDataObject *self)
-    result = (PyTypeObject *)PyType_Type.tp_new(type, args, kwds);
-    if (result == NULL)
-        return NULL;
+    result = (PyTypeObject *)self;
```
```suggestion
```
This and the declaration above can be removed.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2426,62 +2526,51 @@ PyCFuncPtrType_paramfunc(CDataObject *self)
     }
 
-    return (PyObject *)result;
+    Py_CLEAR(attrdict);
```
```suggestion
    Py_DECREF(attrdict);
```

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3253,9 +3386,13 @@ static PPROC FindAddress(void *handle, const char *name, PyObject *type)
+    ctypes_state *st = GLOBAL_STATE();
+    StgInfo *info;
+    if (PyStgInfo_FromType(st, (PyObject *)type, &info) < 0) {
+        return -1;
```
```suggestion
        return NULL;
```

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2609,6 +2698,10 @@ PyCData_traverse(CDataObject *self, visitproc visit, void *arg)
+    PyTypeObject *type = Py_TYPE(self);
+    if (type->tp_flags & Py_TPFLAGS_HEAPTYPE) {
+        Py_VISIT(type);
+    }
```
```suggestion
    Py_VISIT(type);
```
There seems to be no static type left.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2627,8 +2720,15 @@ PyCData_clear(CDataObject *self)
+    PyTypeObject *type = Py_TYPE(self);
+    if (type->tp_flags & Py_TPFLAGS_HAVE_GC) {
+        PyObject_GC_UnTrack(self);
+    }
```
```suggestion
    PyObject_GC_UnTrack(self);
```

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2627,8 +2720,15 @@ PyCData_clear(CDataObject *self)
+    type->tp_free(self);
+    if (type->tp_flags & Py_TPFLAGS_HEAPTYPE) {
+        Py_DECREF(type);
+    }
```
```suggestion
    Py_DECREF(type);
```

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -2773,51 +2892,30 @@ static PyMethodDef PyCData_methods[] = {
+    {0, NULL},
+};
+
+PyType_Spec pycdata_spec = {
```
Not adding `static` specifier to every spec is intentional?

encukou replied:
No, thanks for the catch!

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -4903,48 +4952,27 @@ Simple_repr(CDataObject *self)
+    {Py_tp_new, GenericPyCData_new},
+    {Py_bf_getbuffer, PyCData_NewGetBuffer},
+    {Py_nb_bool, Simple_bool},
+    // traverse, dealloc, clear Py_TPFLAGS_HAVE_GC are inherited
```
I would suggest moving the comment to the base type.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -5248,87 +5304,33 @@ Pointer_subscript(PyObject *myself, PyObject *item)
+static PyType_Slot pycpointer_slots[] = {
+    {Py_tp_doc, PyDoc_STR("XXX to be provided")},
+    {Py_tp_getset, },
+    {Py_tp_getset, Pointer_getsets},
```
```suggestion
    {Py_tp_getset, Pointer_getsets},
```

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -4903,48 +4952,27 @@ Simple_repr(CDataObject *self)
+PyType_Spec pycsimple_spec = {
+    .name = "_ctypes._SimpleCData",
+    .flags = (Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE
+              | Py_TPFLAGS_IMMUTABLETYPE),
```
```suggestion
    .flags = (Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE |
              Py_TPFLAGS_IMMUTABLETYPE),
```

encukou committed:
Inline PyStgInfo_From* & PyStgInfo_Init for performance

encukou committed:
Make CType_Type's clear & dealloc safer

Check that the state and st->PyCType_Type is set.
If they aren't, this will leak rather than crash.

encukou committed:
Remove self decrefs in error cases in init functions

encukou committed:
Add `static` to PyType_Spec

encukou committed:
Apply suggestions from code review

Co-authored-by: neonene <53406459+neonene@users.noreply.github.com>
Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

encukou committed:
Remove redundant comment

encukou reviewed (commented):
> The old StgDict API was designed to be as fast as possible. Could we not just assert() that the result is >= 0 for these calls?

I'd like to think *less* about which functions can currently fail and which can't. Where possible, I'd like the to compiler figure that out for me.
Also, I don't want to micro-optimize without a benchmark. With branch prediction (and especially with LTO), I don't think the `< 0` check is significant. I didn't see a significant change when running all of `test_capi`.

I *did* get a significant speedup by putting these functions in the header, and making them `static inline`. Let's do that. Incidentally, that'll probably also let the compiler figure out that the -1 doesn't happen.

> We've had some segfaults caused by module access in tp_dealloc recently. We should make sure this is really safe.

I made it to leak rather than crash now. Let's leave a proper fix to another PR -- this can't be tested when the state is still global.

encukou committed:
Apply suggestions from code review

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

erlend-aasland commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -459,20 +444,129 @@ static PyType_Spec structparam_spec = {
+        StgInfo *info;
+        if (PyStgInfo_FromType(st, self, &info) < 0) {
+            PyErr_WriteUnraisable(self);
+            PyErr_Clear();
```
Hm, I might be misremembering, but I thought `PyErr_WriteUnraisable` also cleared the raised exception 🤔

erlend-aasland commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -507,104 +601,104 @@ StructUnionType_paramfunc(CDataObject *self)
     /* keep this for bw compatibility */
-    int r = PyDict_Contains(result->tp_dict, &_Py_ID(_abstract_));
+    int r = PyDict_Contains(attrdict, &_Py_ID(_abstract_));
     if (r > 0) {
-        return (PyObject *)result;
+        Py_DECREF(attrdict);
+        return 0;
     }
```
Let's find out if this branch is covered by the test suite.

encukou replied:
It is, by `test.test_ctypes.test_structures.StructureTestCase.test_abstract_class`.

erlend-aasland reviewed (commented):
> I did get a significant speedup by putting these functions in the header, and making them static inline. Let's do that. Incidentally, that'll probably also let the compiler figure out that the -1 doesn't happen.

Yes, putting the PEP-697 wrappers in the header is a neat solution!

I finished an initial read-through of `_ctypes.c`, and it looks good to me. I'll try to find the time to read through it a couple of times more.

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -836,13 +939,14 @@ CDataType_from_param(PyObject *type, PyObject *value)
+            return NULL;
+        }
         /* If we got a PyCArgObject, we must check if the object packed in it
            is an instance of the type's dict->proto */
```
```suggestion
           is an instance of the type's info->proto */
```

neonene reviewed (commented):
Looks good to me, too.  Although `CType_Type_dealloc()` starts without a gc-untrack (against the docs?), I cannot produce a crash through `type.__del__()` when `info->proto` points a ctype class. It seems to me that the known crashes are related to `object.__del__()`: gh-2974, 35e1ff38ee67ee543d9fcb268c3552c5397f9b3f.

encukou committed:
PyErr_WriteUnraisable clears the exception

encukou committed:
Fix comment

Co-authored-by: neonene <53406459+neonene@users.noreply.github.com>

neonene commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -1948,33 +2064,24 @@ PyCSimpleType_paramfunc(CDataObject *self)
     Py_ssize_t proto_len;
     PyMethodDef *ml;
     struct fielddesc *fmt;
 
```
```suggestion

    if (PyType_Type.tp_init(self, args, kwds) < 0) {
        return -1;
    }
```
This can emit `TypeError: type.__init__() takes 1 or 3 arguments`, which would help `CreateSwappedType()` avoid a crash.  An exaggerated version is bc9c8cb74ac9301af16606c99d76c9b432ec7fa1 (#114740).

encukou replied:
Thanks! I also added the test from your PR.

encukou committed:
Call tp_init for argument validation

Co-authored-by: neonene <53406459+neonene@users.noreply.github.com>

neonene committed:
Add test for swapped type creation

encukou committed:
Merge branch 'main' into ctypes-no-stgdict

encukou committed:
Fix test_swapped_type_creation
